### PR TITLE
Add converters to publish metrics as TFMA protobufs

### PR DESCRIPTION
**Pull request by fallonchen**

### CODE REVIEW

File: project/plugins.sbt [hunk 1]
```diff
@@ -7,6 +7,10 @@ addSbtPlugin("org.xerial.sbt" % "sbt-sonatype" % "2.3")
 addSbtPlugin("com.geirsson" % "sbt-scalafmt" % "1.5.1")
 addSbtPlugin("com.typesafe" % "sbt-mima-plugin" % "0.3.0")
 addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.3.4")
-addSbtPlugin("com.eed3si9n" % "sbt-unidoc" % "0.4.2")
```
Comment (regadas) [inline project/plugins.sbt]: nice catch! 👍 😄

File: project/plugins.sbt [hunk 2]
```diff
@@ -7,6 +7,10 @@ addSbtPlugin("org.xerial.sbt" % "sbt-sonatype" % "2.3")
...
 addSbtPlugin("com.typesafe.sbt" % "sbt-ghpages" % "0.6.2")
 addSbtPlugin("com.typesafe.sbt" % "sbt-site" % "1.3.2")
+addSbtPlugin("com.github.gseitz" % "sbt-protobuf" % "0.6.3")
+
+libraryDependencies ++= Seq(
+  "com.github.os72" % "protoc-jar" % "3.3.0.1"
```
Comment (regadas) [inline project/plugins.sbt]: I think we can go with later version like 3.6.0.1?

File: tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+object EvalResult {
+  def apply(metrics: MetricsForSlice): EvalResult = EvalResult(Some(metrics), None)
+  def apply(metrics: MetricsForSlice, plot: Plot): EvalResult =
+    EvalResult(Some(metrics), Some(plot))
+  def apply(plot: Plot): EvalResult = EvalResult(None, Some(plot))
+}
+
+object TfmaConverter {
+
+  implicit val errorRateSummaryConverter
```
Comment (regadas) [inline tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala]: having this implicits here feels a bit off. I would like to purpose a different structure for this. IMO we should avoid implicits in the object scope.

File: tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+                    .newBuilder()
+                    .setDoubleValue(DoubleValue.newBuilder().setValue(value))
+                    .build())
+      .build()
+
+  private def buildDoubleMetrics(metrics: Map[String, Double]): MetricsForSlice = {
+    MetricsForSlice
+      .newBuilder()
+      .setSliceKey(SliceKey.getDefaultInstance)
+      .putAllMetrics(metrics.mapValues { m =>
```
Comment (regadas) [inline tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala]: 🎨 move this outside; more readable?

File: tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  private def buildCalibrationHistogramPlot(ch: List[CalibrationHistogramBucket]): PlotsForSlice =
+    PlotsForSlice
+      .newBuilder()
+      .setSliceKey(SliceKey.getDefaultInstance)
+      .setPlotData(
+        PlotData
+          .newBuilder()
+          .setCalibrationHistogramBuckets(CalibrationHistogramBuckets
+            .newBuilder()
+            .addAllBuckets(ch.map { b =>
```
Comment (regadas) [inline tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala]: 🎨 move outside?

File: tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala (hunk 3 above)
Comment (andrewsmartin) [reply tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala]: +1 `package object extends Implicits` pattern?

File: tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala (hunk 4 above)
Comment (fallonchen) [reply tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala]: Is this what you mean by move this outside?
```
val metrics = metrics.mapValues { m =>
        MetricValue
          .newBuilder()
          .setDoubleValue(
            DoubleValue
              .newBuilder()
              .setValue(m))
          .build()
      }.asJava

    MetricsForSlice
      .newBuilder()
      .setSliceKey(SliceKey.getDefaultInstance)
      .putAllMetrics(metrics)
      .build()
```

File: tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala (hunk 3 above)
Comment (fallonchen) [reply tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala]: I'm familiar with this pattern.. where should I read about it?

File: build.sbt [hunk 6]
```diff
@@ -132,6 +138,35 @@ lazy val docSettings = Def.settings(
...
+      "org.scalanlp" %% "breeze" % breezeVersion,
+      "com.twitter" %% "algebird-core" % algebirdVersion,
+      "org.scalatest" %% "scalatest" % scalaTestVersion,
+      "com.google.protobuf" % "protobuf-java" % protobufVersion % ProtobufConfig.name
+    ),
+    version in ProtobufConfig := protobufVersion,
+    protobufRunProtoc in ProtobufConfig := (args =>
+      // protoc-jar does not include 3.3.1 binary
+      com.github.os72.protocjar.Protoc.runProtoc("-v3.3.0" +: args.toArray)),
+    // Avro and Protobuf files are compiled to src_managed/main/compiled_{avro,protobuf}
```
Comment (regadas) [inline build.sbt]: 🎨 we will not use avro here so i guess we can update/remove this comment 😄

File: tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.twitter.algebird.Aggregator
+import tensorflow_model_analysis.MetricsForSliceOuterClass._
+
+trait TfmaConverter[A, B, T <: Aggregator[A, B, _]] {
+  def convertToTfmaProto(underlying: T): Aggregator[A, B, EvalResult]
+}
+
+sealed trait Plot {
+  val plotData: PlotsForSlice
```
Comment (regadas) [inline tfx/src/main/scala/com/spotify/noether/tfx/TfmaConverter.scala]: @fallonchen 🎨 it's better if you make it `def`

**Comment by fallonchen** [comment]
@regadas I addressed your review comments; can you take a look when you get the chance?

**Pull request merged**